pkg/app/server.go

```diff
@@ -145,21 +145,85 @@ func RunKubeStateMetrics(ctx context.Context, opts *options.Options) error {
 	if config != nil {
-		factories, err = customresourcestate.FromConfig(config)
+		discoveryClient = cgodiscovery.NewDiscoveryClientForConfigOrDie(kubeConfig)
+		// Fill maps, which are required for the store cache invalidation checks to work (in discovery).
+		// We can only resolve GVKs once the discovery has taken place. Think of this as the expensive part of the 2-parter discovery-resolution cache mechanism.
+		gvkMaps.Discover(discoveryClient)
+		// Create CR factories. These basically contain the metric family and metadata.
+		factoriesGetter, err = customresourcestate.ResolveConfig(config, gvkMaps)
 		if err != nil {
-			return fmt.Errorf("Parsing from Custom Resource State Metrics file failed: %v", err)
+			return fmt.Errorf("parsing from Custom Resource State Metrics file failed: %v", err)
+
+		}
+		factories, err = factoriesGetter(nil)
+		if err != nil {
+			return fmt.Errorf("creating Custom Resource State Metrics factories failed: %v", err)
 		}
+
+		// Create stores for the factories and add them to the registry.
+		storeBuilder.WithCustomResourceStoreFactories(factories...)
+		// `availableStores` is initialized with the CR stores and knows how to fetch them now.
+		// We need to do the same thing for any newly added CRs from this point on.
+	}
+
+	// Ensure cache prerequisites are set up only once.
+	var discoverer sync.Once
+	periodicDiscoveryForAppendCRStore := func() {
+		// Start an independent goroutine to periodically update `gvkMaps`.
+		// The interval at which the discovery client will fetch the list of objects.
+		// TODO: Expose as an option?
+		t := time.NewTicker(discovery.Interval)
+		go func() {
+			for range t.C {
+				select {
+				case <-ctx.Done():
+					t.Stop()
+					return
+				default:
+					// Subsequent discoveries will only occur if a discovery was previously made.
+					gvkMaps.Discover(discoveryClient)
+					if gvkMaps.UpdateStores {
+						klog.Infoln("maps differ, updating CR stores")
+						// Get the new GVKs between r.Map and r.OldMap.
+						var resources []customresourcestate.Resource
+						for _, gvk := range gvkMaps.MapDifference {
+							resource := customresourcestate.Resource{}
+							resource.GroupVersionKind = customresourcestate.GroupVersionKind(gvk)
+							resources = append(resources, resource)
+						}
+						// Get the new factories.
+						customFactories, err := factoriesGetter(resources)
+						if err != nil {
+							klog.ErrorS(err, "discovery ticker: creating Custom Resource State Metrics factories failed")
+						}
+						// Add the new factories to the store builder.
+						storeBuilder.WithCustomResourceStoreFactories(customFactories...)
+						// Update the registry with the new stores.
+						storeBuilder.WithGenerateCustomResourceStoresFunc(storeBuilder.DefaultGenerateCustomResourceStoresFunc())
+						// Reset the flag, if there were no errors. Else, we'll try again on the next tick.
+						gvkMaps.UpdateStores = false
```

Comment:
Are you sure you don't need a mutex to update this? This particular invocation is in a goroutine.

Reply:
Are you running tests with data-race enabled?

Reply:
> Are you sure you don't need a mutex to update this? This particular invocation is in a goroutine.

Hmm, right now only one goroutine will be running at any given time (owing to `sync.Once` block around the caller) and operations are synchronous, but someone setting the value in the future outside of this block will need to take care of race-conditions, I'll make that change! 👍🏼 

> Are you running tests with data-race enabled?

I haven't wrote e2e tests for this PR yet, just some initial unit ones, but I'll make sure to use that flag once I do.

Reply:
Added.